modules/tensor_mechanics/src/materials/CalculateStrainLagrangianKernel.C

```diff
@@ -0,0 +1,248 @@
+  {
+    _disp[i] = &coupledValue("displacements", i);
+    _grad_disp[i] = &coupledGradient("displacements", i);
+  }
+
+  // All others zero (so this will work naturally for plane strain problems)
+  for (unsigned int i = _ndisp; i < 3; i++)
+  {
+    _disp[i] = &_zero;
+    _grad_disp[i] = &_grad_zero;
+  }
+}
+
+void
+CalculateStrainLagrangianKernel::initQpStatefulProperties()
+{
+  _total_strain[_qp].zero();
+  _mechanical_strain[_qp].zero();
+  _strain_increment[_qp].zero();
+  _rotation_increment[_qp] = RankTwoTensor::Identity();
+  _def_grad[_qp] = RankTwoTensor::Identity();
+  _df[_qp] = RankTwoTensor::Identity();
+  _inv_def_grad[_qp] = RankTwoTensor::Identity();
+  _detJ[_qp] = 1.0;
+  _homogenization_contribution[_qp].zero();
+}
+
+void
+CalculateStrainLagrangianKernel::computeQpProperties()
+{
+  // We'll calculate the deformation gradient even for small kinematics
+  _def_grad[_qp] =
+      (RankTwoTensor::Identity() +
+       RankTwoTensor((*_grad_disp[0])[_qp], (*_grad_disp[1])[_qp], (*_grad_disp[2])[_qp]));
+
+  // Add in the macorscale gradient contribution
+  _homogenization_contribution[_qp] = _homogenizationContribution();
+  _def_grad[_qp] += _homogenization_contribution[_qp];
+
+  // If the kernel is large deformation then we need the "actual"
+  // kinematic quantities
+  RankTwoTensor L;
+  if (_ld_kernel)
+  {
+    _inv_def_grad[_qp] = _def_grad[_qp].inverse();
+    _detJ[_qp] = _def_grad[_qp].det();
+    _df[_qp] = _def_grad_old[_qp] * _inv_def_grad[_qp];
+    L = RankTwoTensor::Identity() - _df[_qp];
+  }
+  // For small deformations we just provide the identity
+  else
+  {
+    _inv_def_grad[_qp] = RankTwoTensor::Identity();
+    _detJ[_qp] = 1.0;
+    _df[_qp] = RankTwoTensor::Identity();
+    L = _def_grad[_qp] - _def_grad_old[_qp];
+  }
+
+  // This is a useful case, but I need to figure out what these MOOSE
+  // models need and return
+  if (_ld_kernel && _ld_material)
+  {
+    mooseError("Strain calculations for material large displacements not"
+               " implemented");
+  }
+  // This is for the objective update
+  else if (_ld_kernel && !_ld_material)
+  {
+    // Supply the deformation rate as the increment in strain
+    _calculateIncrementalStrains(L);
+  }
+  // This case makes no sense
+  else if (!_ld_kernel && _ld_material)
+  {
+    mooseError("Lagrangian kernel small deformations with MOOSE material large"
+               " deformations is not implemented");
+  }
+  // Small strains
+  else
+  {
+    // Need total_strain and mechanical_strain as small strains
+    _calculateIncrementalStrains(L);
+  }
+}
+
+void
+CalculateStrainLagrangianKernel::_calculateIncrementalStrains(RankTwoTensor L)
+{
+  // Get the deformation increments
+  RankTwoTensor D = (L + L.transpose()) / 2.0;
+  _strain_increment[_qp] = D - _eigenstrainIncrement();
+
+  // Don't do it MOOSE! The WrapStress object will take care of it
+  _rotation_increment[_qp] = RankTwoTensor::Identity();
```

Comment:
This worth some discussion I think. You are basically forcing the existing stress calculators in MOOSE to "skip" the rotation from the intermediate configuration to the current configuration, and doing that rotation in the "wrapper".